client/state/themes/actions.js

```diff
@@ -454,3 +460,83 @@ export function clearThemeUpload( siteId ) {
 		siteId,
 	};
 }
+
+export function initiateThemeTransfer( siteId, file ) {
+	return dispatch => {
+		dispatch( {
+			type: THEME_TRANSFER_INITIATE_REQUEST,
+			siteId,
+		} );
+		return wpcom.undocumented().initiateTransfer( siteId, null, file, ( event ) => {
+			dispatch( {
+				type: THEME_TRANSFER_INITIATE_PROGRESS,
+				siteId,
+				loaded: event.loaded,
+				total: event.total,
+			} );
+		} )
+			.then( ( { transfer_id } ) => {
+				dispatch( {
+					type: THEME_TRANSFER_INITIATE_SUCCESS,
+					siteId,
+					transferId: transfer_id,
+				} );
+				themeTransferStatus( siteId, transfer_id )( dispatch );
+			} )
+			.catch( error => {
+				dispatch( {
+					type: THEME_TRANSFER_INITIATE_FAILURE,
+					siteId,
+					error,
+				} );
+			} );
+	};
+}
+
+export function themeTransferStatus( siteId, transferId, interval = 3000, timeout = 180000 ) {
+	const endTime = Date.now() + timeout;
+	return dispatch => {
+		const pollStatus = ( resolve, reject ) => {
+			if ( Date.now() > endTime ) {
+				// timed-out, stop polling
+				dispatch( {
+					type: THEME_TRANSFER_STATUS_FAILURE,
+					siteId,
+					transferId,
+					error: 'client timeout',
+				} );
+				return resolve();
+			}
+			return wpcom.undocumented().transferStatus( siteId, transferId )
+				.then( ( { status, message, themeId } ) => {
+					dispatch( {
+						type: THEME_TRANSFER_STATUS_SUCCESS,
```

Comment:
Yeah, this _is_ slightly confusing, but does match up with the way the API works, I think.

Only the /initiate endpoint has a `PROGRESS` action, which is an update from the browser specifying how much of the zip has been uploaded.

The /status endpoint, which we are dealing with here, has only `SUCCESS`, which will have some associated status (not necessarily _complete_) and `FAILURE`, which is the error case.